### PR TITLE
EUDEMO upper port size optimisation problem

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Starting to address #906 

## Description
Adds an optimisation problem for the sizing of the upper port and the segmentation of the breeding blankets in the poloidal plane.

## Interface Changes
None yet, still need to figure out how best to thread this into the reactor build chain. Thought I'd get the brains working first.

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/builders/EUDEMO/maintenance.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        constraints = [
+            OptimisationConstraint(
+                self.constrain_blanket_cut,
+                f_constraint_args={
+                    "bb": breeding_blanket_xz,
+                    "c_rm": c_rm,
+                    "r_ib_min": r_ib_min,
+                    "r_ob_max": r_ob_max,
+                },
+                tolerance=1e-6 * np.ones(3),
```
Comment (hsaunders1904) [inline bluemira/builders/EUDEMO/maintenance.py:90]: I'm not clear where this 3 comes from, I guess it's the number of constraints? Maybe another one to extract into a variable?

**Review by hsaunders1904 — CHANGES_REQUESTED** [review]
Code looks pretty good to me. Only a few nitpicks, we're also missing tests.

File: bluemira/builders/EUDEMO/maintenance.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    """
+    Make the void geometry for the upper port in the poloidal plane.
+    """
+    x = [r_up_inner, r_up_outer, r_up_outer, r_up_inner]
+    z = [z_min, z_min, z_max, z_max]
+    return BluemiraFace(make_polygon({"x": x, "y": 0, "z": z}, closed=True))
+
+
+if __name__ == "__main__":
+    # TODO: Remove this once threaded into the reactor build via some kind of builder
```
Comment (hsaunders1904) [inline bluemira/builders/EUDEMO/maintenance.py:248]: Do we have an issue that resolves this TODO? If there's an existing issue, is it explicit about removing this?

**Review by hsaunders1904 — COMMENTED** [review]
A couple of minor comments to finish

File: bluemira/builders/EUDEMO/maintenance.py [hunk 3]
```diff
@@ -238,6 +238,22 @@ def segment_blanket_xz(breeding_blanket_xz, r_inner_cut, cut_angle, cut_thicknes
 def build_upper_port_zone(r_up_inner, r_up_outer, z_max=10, z_min=0):
     """
     Make the void geometry for the upper port in the poloidal plane.
+
+    Parameters
+    ----------
+    r_up_inner: radius
```
Comment (hsaunders1904) [inline bluemira/builders/EUDEMO/maintenance.py:244]: Is `radius` a type, or should this be `float`? 

Happy for this to be merged once this is fixed

**Review by hsaunders1904 — APPROVED** [review]

File: bluemira/builders/EUDEMO/maintenance.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        constraints = [
+            OptimisationConstraint(
+                self.constrain_blanket_cut,
+                f_constraint_args={
+                    "bb": breeding_blanket_xz,
+                    "c_rm": c_rm,
+                    "r_ib_min": r_ib_min,
+                    "r_ob_max": r_ob_max,
+                },
+                tolerance=1e-6 * np.ones(n_constraints),
```
Comment (je-cook) [inline bluemira/builders/EUDEMO/maintenance.py:96]: Is it overkill to make this tolerance configurable? Or maybe just 0.1*geometry.constants.MINIMUM_LENGTH

**Review by je-cook — COMMENTED** [review]
Just one minor from me

**Review by je-cook — APPROVED** [review]

**Pull request merged**